iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java

```diff
@@ -2095,6 +2097,55 @@ public List<Partition> getPartitionsByExpr(org.apache.hadoop.hive.ql.metadata.Ta
     }
   }
 
+  @Override
+  public List<Partition> getPartitionsWithFilter(org.apache.hadoop.hive.ql.metadata.Table hmsTable,
+      ExprNodeDesc filter, boolean currentSpec) {
+    SearchArgument sarg = ConvertAstToSearchArg.create(conf, (ExprNodeGenericFuncDesc) filter);
+    Expression exp = HiveIcebergFilterFactory.generateFilterExpression(sarg);
+    Table table = IcebergTableUtil.getTable(conf, hmsTable.getTTable());
+    int tableSpecId = table.spec().specId();
+    List<Partition> partitions = Lists.newArrayList();
+
+    TableScan scan = table.newScan().filter(exp).caseSensitive(false).includeColumnStats().ignoreResiduals();
+
+    try (CloseableIterable<FileScanTask> tasks = scan.planFiles()) {
+      tasks.forEach(task -> {
+        DataFile file = task.file();
+        PartitionSpec spec = task.spec();
+        if ((currentSpec && file.specId() == tableSpecId) || (!currentSpec && file.specId() != tableSpecId)) {
```

Comment:
`file.specId() == table.spec().specId() || !latestSpecOnly`

Reply:
I want this method to return all files for the latest spec when latestSpecOnly=True, or all files with older spec when  latestSpecOnly=False. If we change the condition to the one above, it will always return files for the latest spec regardless of latestSpecOnly value.

Reply:
is there a bug then in IcebergTableUtil#getPartitionInfo? and we should change to
````
latestSpecOnly? file.specId() == table.spec().specId() : file.specId() != table.spec().specId()
````

Reply:
Not a bug, I'm using `IcebergTableUtil#getPartitionInfo` to return partitions for all specs or for the latest spec only, and `IcebergTableUtil#getPartitionsByExpr` returns partitions with the latest spec or partitions with old specs.

Reply:
Done, added Javadoc